// videojs_snapshot_new/scripts/modules.rollup.config.js
/*
 * @Author: your name
 * @Date: 2020-01-13 16:56:19
 * @LastEditTime: 2020-01-13 17:35:53
 * @LastEditors: your name
 * @Description: In User Settings Edit
 * @FilePath: \video-js-snapshot-master\scripts\modules.rollup.config.js
 */
/**
 * Rollup configuration for packaging the plugin in a module that is consumable
 * by either CommonJS (e.g. Node or Browserify) or ECMAScript (e.g. Rollup).
 *
 * These modules DO NOT include their dependencies as we expect those to be
 * handled by the module system.
 */
import babel from 'rollup-plugin-babel';
import json from 'rollup-plugin-json';

export default {
    moduleName: 'videojsSnapshot',
    entry: 'src/js/videojs-snapshot.js',
    external: [
        'global',
        'global/document',
        'global/window',
        'video.js'
    ],
    globals: {
        'video.js': 'videojs'
    },
    legacy: true,
    plugins: [
        json(),
        babel({
            babelrc: false,
            exclude: 'node_modules/**',
            presets: [
                'es3', ['es2015', {
                    loose: true,
                    modules: false
                }]
            ],
            plugins: [
                'external-helpers',
                'transform-object-assign'
            ]
        })
    ],
    targets: [
        { dest: 'dist/videojs-snapshot.cjs.js', format: 'cjs' },
        { dest: 'dist/videojs-snapshot.es.js', format: 'es' }
    ]
};
